File: isegoria/src/main/java/com/eulersbridge/isegoria/network/LenientLongAdapter.java
package com.eulersbridge.isegoria.network;

import android.support.annotation.Nullable;

import com.squareup.moshi.FromJson;
import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.JsonReader;
import com.squareup.moshi.JsonWriter;
import com.squareup.moshi.ToJson;

import java.io.IOException;

/**
 * Needed to handle cases of null IDs for users/contacts/institutions/etc.
 */
class LenientLongAdapter extends JsonAdapter<Long> {

    @FromJson
    public Long fromJson(JsonReader reader) throws IOException {

        try {
            JsonReader.Token token = reader.peek();

            if (token == JsonReader.Token.NULL) {
                reader.nextNull();
                return 0L;

            } else {
                return reader.nextLong();
            }
        } catch (Exception e) {
            return 0L;
        }
    }

    @ToJson
    public void toJson(JsonWriter writer, @Nullable Long value) throws IOException {
        writer.value(value);
    }
}

File: isegoria/src/main/kotlin/com/eulersbridge/isegoria/network/adapters/LenientLongAdapter.kt
package com.eulersbridge.isegoria.network.adapters

import com.squareup.moshi.*
import java.io.IOException

/**
 * Needed to handle cases of null IDs for users/contacts/institutions/etc.
 */
class LenientLongAdapter : JsonAdapter<Long>() {

    @FromJson
    @Throws(IOException::class)
    override fun fromJson(reader: JsonReader): Long? {

        return try {
            val token = reader.peek()

            if (token == JsonReader.Token.NULL) {
                reader.nextNull<Any>()
                0L

            } else {
                reader.nextLong()
            }
        } catch (e: Exception) {
            0L
        }
    }

    @ToJson
    @Throws(IOException::class)
    override fun toJson(writer: JsonWriter, value: Long?) {
        writer.value(value)
    }
}